File: leetcode/ruby/687_longest_univalue_path.rb
def longest_univalue_path(root)
  @longest = 0
  def longest_univalue_path_helper(node)
  return 0 unless node
  left_route = longest_univalue_path_helper(node.left)
  right_route = longest_univalue_path_helper(node.right)
  if node.left && node.left.val == node.val
    left = left_route + 1
  else
    left = 0
  end

  if node.right && node.right.val == node.val
    right = right_route + 1
  else
    right = 0
  end
  @longest = [@longest, left + right].max
  [left, right].max
  end

  longest_univalue_path_helper(root)
  @longest
end

File: leetcode/python/687_longest_univalue_path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        self.longest = 0

        def traverse(node, parent_val):
            if not node:
                return 0
            left, right = traverse(node.left, node.val), traverse(
                node.right, node.val)
            self.longest = max(self.longest, left + right)
            return 1 + max(left, right) if node.val == parent_val else 0
        traverse(root, None)
        return self.longest
